File: bot/pkg/claude/protocol.go
package claude

// Go → Bridge messages (sent to bridge stdin)

// BridgeStartMsg tells the bridge to initialize for a project
type BridgeStartMsg struct {
	Type           string `json:"type"`            // "start"
	ProjectPath    string `json:"project_path"`
	PermissionMode string `json:"permission_mode"` // default, bypassPermissions, acceptEdits, plan
	SessionID      string `json:"session_id,omitempty"`
}

// BridgeUserMsg sends a user message to the bridge
type BridgeUserMsg struct {
	Type    string `json:"type"` // "user_message"
	Content string `json:"content"`
}

// BridgeToolResponseMsg responds to a tool_request/ask_user/plan_review from the bridge
type BridgeToolResponseMsg struct {
	Type      string          `json:"type"`       // "tool_response"
	RequestID string          `json:"request_id"`
	Result    BridgeToolResult `json:"result"`
}

// BridgeToolResult is the allow/deny response for a tool request
type BridgeToolResult struct {
	Behavior     string                 `json:"behavior"` // "allow" or "deny"
	UpdatedInput map[string]interface{} `json:"updated_input,omitempty"`
	Message      string                 `json:"message,omitempty"`
}

// BridgeInterruptMsg tells the bridge to stop the current execution
type BridgeInterruptMsg struct {
	Type string `json:"type"` // "interrupt"
}

// Bridge → Go messages (received from bridge stdout)

// BridgeMessage is the union type for all messages from the bridge
type BridgeMessage struct {
	Type      string `json:"type"`
	SessionID string `json:"session_id,omitempty"`

	// assistant_text
	Content string `json:"content,omitempty"`

	// tool_request
	RequestID string                 `json:"request_id,omitempty"`
	ToolName  string                 `json:"tool_name,omitempty"`
	Input     map[string]interface{} `json:"input,omitempty"`

	// ask_user
	Questions []BridgeQuestion `json:"questions,omitempty"`

	// plan_review
	Plan string `json:"plan,omitempty"`

	// result
	Status     string  `json:"status,omitempty"`
	Result     string  `json:"result,omitempty"`
	CostUSD    float64 `json:"cost_usd,omitempty"`
	DurationMs int64   `json:"duration_ms,omitempty"`

	// error
	Message string `json:"message,omitempty"`
}

// BridgeQuestion represents a question from AskUserQuestion
type BridgeQuestion struct {
	Question    string                `json:"question"`
	Header      string                `json:"header"`
	Options     []BridgeQuestionOption `json:"options"`
	MultiSelect bool                  `json:"multi_select"`
}

// BridgeQuestionOption represents one option in an AskUserQuestion
type BridgeQuestionOption struct {
	Label       string `json:"label"`
	Description string `json:"description"`
}

File: bot/bridge/src/protocol.ts
// Go → Bridge messages (received on stdin)
export type GoMessage =
  | StartMessage
  | UserMessage
  | ToolResponseMessage
  | InterruptMessage;

export interface StartMessage {
  type: "start";
  project_path: string;
  permission_mode: "default" | "bypassPermissions" | "acceptEdits" | "plan";
  session_id?: string; // resume existing session
}

export interface UserMessage {
  type: "user_message";
  content: string;
}

export interface ToolResponseMessage {
  type: "tool_response";
  request_id: string;
  result: ToolResult;
}

export type ToolResult =
  | { behavior: "allow"; updated_input: Record<string, unknown> }
  | { behavior: "deny"; message: string };

export interface InterruptMessage {
  type: "interrupt";
}

// Bridge → Go messages (sent on stdout)
export type BridgeMessage =
  | InitMessage
  | AssistantTextMessage
  | ToolRequestMessage
  | AskUserMessage
  | PlanReviewMessage
  | ResultMessage
  | ErrorMessage;

export interface InitMessage {
  type: "init";
  session_id: string;
}

export interface AssistantTextMessage {
  type: "assistant_text";
  content: string;
}

export interface ToolRequestMessage {
  type: "tool_request";
  request_id: string;
  tool_name: string;
  input: Record<string, unknown>;
}

export interface AskUserMessage {
  type: "ask_user";
  request_id: string;
  questions: AskUserQuestion[];
}

export interface AskUserQuestion {
  question: string;
  header: string;
  options: { label: string; description: string }[];
  multi_select: boolean;
}

export interface PlanReviewMessage {
  type: "plan_review";
  request_id: string;
  plan: string;
}

export interface ResultMessage {
  type: "result";
  status: "success" | "error";
  result: string;
  cost_usd: number;
  session_id: string;
  duration_ms: number;
}

export interface ErrorMessage {
  type: "error";
  message: string;
}
